py/desiutil/annotate.py

```diff
@@ -0,0 +1,376 @@
+    """Parse a YAML file that contains column names and units and optionally comments.
+
+    The YAML file should contain a dictionary with a keyword like 'units' and,
+    optionally, a keyword like 'comments'.
+
+    For backwards-compatibility, the YAML file can be simply a dictionary
+    containing column names.
+
+    Parameters
+    ----------
+    filename : :class:`str` or :class:`pathlib.Path`
+        Read column definitions from `filename`.
+
+    Returns
+    -------
+    :class:`tuple`
+        A tuple containing two :class:`dict` objects for units and comments.
+        If no comments are detected, the comments :class:`dict` will be empty.
+    """
+    comments = dict()
+    # log.debug("y = yaml.safe_load('%s')", filename)
+    with open(filename, newline='') as f:
+        y = yaml.safe_load(f)
+    try:
+        u = yml_unit_key(y)
+    except KeyError:
+        log.warning(f"{filename} does not have a unit column, assuming keys are columns!")
+        u = None
+    try:
+        c = yml_unit_key(y, comment=True)
+    except KeyError:
+        c = None
+    if u:
+        units = y[u]
+    else:
+        units = y
+    if c:
+        comments = y[c]
+    return (units, comments)
+
+
+def annotate_table(table, units, comments=None, inplace=False):
+    """Add annotations to `table`.
+
+    Parameters
+    ----------
+    table : :class:`astropy.table.Table` or :class:`astropy.table.QTable`
+        A data table.
+    units : :class:`dict`
+        Mapping of table columns to units.
+    comments : :class:`dict`, optional
+        Mapping of table columns to comments.
+    inplace : :class:`bool`, optional
+        If ``True``, modify `table` directly instead of returning a copy.
+
+    Returns
+    -------
+    :class:`astropy.table.Table`
+        An updated version of `table`.
+    """
+    if inplace:
+        t = table
+    else:
+        if isinstance(table, QTable):
+            t = QTable(table)  # copy=True is the default.
+        else:
+            t = Table(table)
+    for column in units:
+        if column in t.colnames:
+            if len(units[column]) > 0:
+                try:
+                    log.debug("t['%s'].unit = '%s'", column, units[column])
+                    t[column].unit = units[column]
+                except AttributeError:
+                    #
+                    # Can't change .unit if it is already set. Try to convert.
+                    #
+                    try:
+                        log.debug("t.replace_column('%s', t['%s'].to('%s'))", column, column, units[column])
+                        t.replace_column(column, t[column].to(units[column]))
+                    except UnitConversionError:
+                        log.error("Cannot add or replace unit '%s' to column '%s'!", units[column], column)
+            else:
+                log.info("Not setting blank unit for column '%s'.", column)
+        else:
+            log.info("Column '%s' not present in table.", column)
```

Comment:
let's downgrade this to a `log.debug`, since the likely common case of using desidatamodel `column_descriptions.csv` will have N>>1 columns not in any given table, which generates a lot of logging cruft.

On the flip side, let's explicitly log if the table has columns that are *not* in `units`.  That's not an error case, but I think is worth logging.  Something like
```python
for colname in t.colnames:
    if colname not in units:
        log.info('Column %s not found in input units table', colname)
```

Reply:
Yes, that's fine.